File: src/main/java/leetcode/linked_list/lru_cache/LRUCache.java
package leetcode.linked_list.lru_cache;

import java.util.HashMap;
import java.util.Map;

/**
 * @ClassName LRUCache
 * @Description sentinel/dummy tail and head node is very useful
 * @Author cy
 * @Date 2021/5/9 17:00
 */
public class LRUCache {

    int capacity;

    Map<Integer, Node> map = new HashMap<>();

    Node head;

    Node tail;


    static class Node {
        int key;

        int value;

        Node prev;

        Node next;

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.head = new Node(0,0);
        this.tail = new Node(0,0);
        this.head.next = tail;
        this.tail.prev = head;
    }

    public int get(int key) {
        if(map.containsKey(key))
        {
            Node node = map.get(key);
            remove(node);
            insert(node);
            return node.value;
        }
        else
        {
            return -1;
        }
    }

    /**
     * Update the value of the key if the key exists.
     * Otherwise, add the key-value pair to the cache.
     * If the number of keys exceeds the capacity from this operation, evict the least recently used key
     * @param key
     * @param value
     */
    public void put(int key, int value) {
        if(map.containsKey(key))
        {
            remove(map.get(key));
        }
        if(map.size() == capacity)
        {
            remove(tail.prev);
        }
        insert(new Node(key,value));
    }

    private void insert(Node node) {
        map.put(node.key,node);
        Node headNext = head.next;
        headNext.prev = node;
        node.next = headNext;
        head.next = node;
        node.prev = head;
    }

    private void remove(Node node) {
        map.remove(node.key);
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

}

File: leetcode/146-lru_cache/LRU Cache.go
package leetcode

type LRUCache struct {
	Capacity 	int
	Keys 		map[int]*Node
	// head is the most recently used node
	// tail is the least recently used node
	head, tail 	*Node
}

type Node struct {
	Key, Value	int
	// doubly linked list
	Prev, Next	*Node
}


func Constructor(capacity int) LRUCache {
    return LRUCache{
	    Keys: 	make(map[int]*Node),
	    Capacity: 	capacity,
    }
}


func (this *LRUCache) Get(key int) int {
    if node, ok := this.Keys[key]; ok {
	    this.Remove(node) // remove from current position
	    this.Add(node) // add to head
	    return node.Value
    }

    return -1
}


func (this *LRUCache) Put(key int, value int)  {
    if node, ok := this.Keys[key]; ok {
	    node.Value = value
	    this.Remove(node)
	    this.Add(node)
	    return
    }else {
	    node := &Node{ Key: key, Value: value }
	    this.Keys[key] = node
	    this.Add(node)
    }

    if len(this.Keys) > this.Capacity {
	    delete(this.Keys, this.tail.Key)
	    this.Remove(this.tail)
    }
}


func (this *LRUCache) Add(node *Node) {
	node.Next = this.head
	node.Prev = nil
	if this.head != nil {
		this.head.Prev = node
	}

	this.head = node

	if this.tail == nil {
		this.tail = node
		this.tail.Next = nil
	}
}

func (this *LRUCache) Remove(node *Node) {
	if node == this.head {
		this.head = node.Next
		node.Next = nil
		return
	}
	if node == this.tail {
		this.tail = node.Prev
		node.Prev.Next = nil
		node.Prev = nil
		return
	}
	// remove node in LinkedList
	node.Prev.Next = node.Next
	node.Next.Prev = node.Prev
}

